### PR TITLE
Windows interop

### PR DESCRIPTION
**Pull request by richardschneider**
See the discussion at https://github.com/ipfs/js-ipfs-api/issues/623

### CODE REVIEW

**Comment by diasdavid** [comment]
Pulling discussion from https://github.com/ipfs/js-ipfs-api/issues/623#issuecomment-343772425

------------

@richardschneider  mind opening a PR on js-ipfs-api?

> diag.sys [Error: this operation is not supported on your platform]

We can have a flag that disables this test for Windows, it's ok.

> dial attempt failed: context deadline exceeded

Do you know what it was trying to dial to?

> .swarm.addrs Uncaught AssertionError: expected [] not to be empty

Sounds like this failed due to the previous dial fail

> .swarm.peers timeout

Increase the timeout?

> various issues with addFromFs

Reported here https://github.com/ipfs/js-ipfs-api/issues/408

> a few tests are getting different hash IDs

Weird, which ones?

**Comment by diasdavid** [comment]
> Test output from @richardschneider 

https://github.com/ipfs/js-ipfs-api/files/1465164/test-results.txt

**Comment by thisconnect** [comment]
There are a few slashes `'/'` in get-files-stream.js https://github.com/ipfs/js-ipfs-api/blob/master/src/utils/get-files-stream.js#L49 and possible in other places. Not sure if they matter but could you try with `path.sep` https://nodejs.org/api/path.html#path_path_sep ?

**Comment by richardschneider** [comment]
@thisconnect Thanks.  I plan to work on this code today.  There are two issues:
- the file name is `resolved` so will have os-specific separator
- `glob` wants forward slashes

**Comment by richardschneider** [comment]
On my local windows I'm getting ECONNREFUSED in https://github.com/ipfs/js-ipfs-api/blob/master/test/ipfs-factory/client.js

It is trying to talk to localhost:55155
Who is listening to this port? And why is it hardcoded?

**Comment by richardschneider** [comment]
Solved the port 55155 issue. I was using `mocha` directly to only run a few tests.  The tests do NOT spin up a test server.  That is aegir responsibility, so I didn't have a test server.  Better error messages could have saved me a few hours.

**Comment by richardschneider** [comment]
@diasdavid I have made the minor changes to get this running on Windows.

All the red lights have nothing to do with Windows and are pre-existing problems.

File: src/utils/request-api.js [hunk 1]
```diff
@@ -57,7 +57,15 @@ function onRes (buffer, cb) {
...
-          err = JSON.parse(err)
+          // Not all errors are JSON
+          try {
+            err = JSON.parse(err)
+          } catch (e) {
+            err = {
+              Code: 'n/a',
+              Message: err
+            }
+          }
```
Comment (diasdavid) [inline src/utils/request-api.js]: 👍

File: test/diag.spec.js [hunk 2]
```diff
@@ -6,10 +6,16 @@ const chai = require('chai')
...
+const os = require('os')
 
 describe('.diag', function () {
   this.timeout(50 * 1000)
 
+  if (os.platform() === 'win32') {
+    it('skip these on Windows')
+    return
+  }
+
```
Comment (diasdavid) [inline test/diag.spec.js]: 👍

**Comment by richardschneider** [comment]
The code can NOT handle x-stream-error correctly.  See [request-api](https://github.com/ipfs/js-ipfs-api/blob/e2d894c00be9508d6373c144da9796ececf59a12/src/utils/request-api.js#L51)

>// TODO: This needs reworking.
      // this is a chicken and egg problem -
      // 1) we can't get Trailer headers unless the response ends
      // 2) we can't propagate the error, because the response stream
      // is closed
      // (perhaps we can workaround this using pull-streams)

**Pull request merged**